### PR TITLE
[SPARK-48035][SQL] Fix try_add/try_multiply being semantic equal to add/multiply

### PR DESCRIPTION
**Pull request by db-scnakandala**
### What changes were proposed in this pull request?
- This PR fixes a correctness bug in commutative operator canonicalization where we currently do not take into account the evaluation mode during operand reordering. 
- As a result, the following condition will be incorrectly true:
```
val l1 = Literal(1)
val l2 = Literal(2)
val l3 = Literal(3)
val expr1 = Add(Add(l1, l2), l3)
val expr2 = Add(Add(l2, l1, EvalMode.TRY), l3)
expr1.semanticEquals(expr2) 
```
- To fix the issue, we now reorder commutative operands only if all operators have the same evaluation mode.
 
### Why are the changes needed?
- To fix a correctness bug.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Added unit tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by db-scnakandala** [comment]
cc: @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala [hunk 1]
```diff
@@ -454,4 +454,29 @@ class CanonicalizeSuite extends SparkFunSuite {
     // different.
     assert(common3.canonicalized != common4.canonicalized)
   }
+
+  test("[SPARK-48035] Add/Multiply operator canonicalization should take into account the" +
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala:458]:
```suggestion
  test("SPARK-48035: Add/Multiply operator canonicalization should take into account the" +
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 2]
```diff
@@ -452,13 +452,14 @@ case class Add(
     copy(left = newLeft, right = newRight)
 
   override lazy val canonicalized: Expression = {
-    // TODO: do not reorder consecutive `Add`s with different `evalMode`
-    val reorderResult = buildCanonicalizedPlan(
+    val evalModes = collectEvalModes(this, {case Add(_, _, evalMode) => Seq(evalMode)})
+    lazy val reorderResult = buildCanonicalizedPlan(
       { case Add(l, r, _) => Seq(l, r) },
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:457]: shall we simply add check here? `case Add(l, r, em) if em == evalMode`

Comment (db-scnakandala) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:457]: That is neat! I will create a follow-up PR.

**Pull request closed without merging**